# Question 1, 2, 3 - Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto

# Question1

In [5]:
# importing data
import pandas as pd
import requests
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [6]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [7]:
len(wiki_data), type(wiki_data)

(3, list)

We need the first table alone, so dropping the other tables

In [8]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Dropping Borough which are not assigned

In [9]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [10]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [11]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [12]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [13]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [14]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 column

# Question 2 of Week 3 - Capstone - Include latitudes and longitudes

Alternatively, as suggested in the assignment, Importing the CSV file from the URL

In [15]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [16]:
print("the shape of our csv data is: ", data.shape)

the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [17]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [18]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [19]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [20]:
combined_data.shape

(103, 5)

Solution: We get 103 rows as expected when we do a inner join, so we have good data.

# Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [21]:
from geopy.geocoders import Nominatim
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [22]:
!pip install folium
import folium

     |████████████████████████████████| 94 kB 3.1 MB/s eta 0:00:011


In [23]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

In [24]:
CLIENT_ID = 'ICD30E1A4NTTD2FS3ACVWQRZ5XUEG034M14SXLSU2EFCVZNO' 
CLIENT_SECRET = 'JMWI0MEEUITOKQRQTGGCON45RSTJCFJN0A2Y2SDGOWYV3W5A'
VERSION = '20211101' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' , CLIENT_ID)
print('CLIENT_SECRET:' , CLIENT_SECRET)

Your credentails:
CLIENT_ID:  ICD30E1A4NTTD2FS3ACVWQRZ5XUEG034M14SXLSU2EFCVZNO
CLIENT_SECRET: JMWI0MEEUITOKQRQTGGCON45RSTJCFJN0A2Y2SDGOWYV3W5A


Next, we create a function to get all the venue categories in Toronto

In [37]:
neighborhood_latitude = combined_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = combined_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = combined_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


In [38]:
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=ICD30E1A4NTTD2FS3ACVWQRZ5XUEG034M14SXLSU2EFCVZNO&client_secret=JMWI0MEEUITOKQRQTGGCON45RSTJCFJN0A2Y2SDGOWYV3W5A&v=20211101&ll=43.7532586,-79.3296565&radius=500&limit=100'

Collecting the venues in Toronto for each Neighbourhood

In [39]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ffd8e989fc7d574b2a3f99c'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 3,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

In [28]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [29]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:6: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1  Brookbanks Pool               Pool  43.751389 -79.332184
2    Variety Store  Food & Drink Shop  43.751974 -79.333114

In [30]:

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [41]:

toronto_venues = getNearbyVenues(names=combined_data['Neighbourhood'],
                                   latitudes=combined_data['Latitude'],
                                   longitudes=combined_data['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [42]:
print(toronto_venues.shape)
toronto_venues.head()

(2122, 7)


Neighbourhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1         Brookbanks Pool       43.751389       -79.332184   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1                   Pool  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant

Checking how many venues there are for each venue

In [43]:
toronto_venues.groupby('Neighbourhood').count()

Neighborhood Latitude  \
Neighbourhood                                                            
Agincourt                                                            4   
Alderwood, Long Branch                                               7   
Bathurst Manor, Wilson Heights, Downsview North                     22   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   21   
...                                                                ...   
Willowdale, Willowdale East                                         34   
Willowdale, Willowdale West                                          4   
Woburn                                                               4   
Woodbine Heights                                                     7   
York Mills West                                                      3   

                                                 Neighborhood Longitude  \
Neighbourhood                                                             
Agincourt                                                             4   
Alderwood, Long Branch                                                7   
Bathurst Manor, Wilson Heights, Downsview North                      22   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    21   
...                                                                 ...   
Willowdale, Willowdale East                                          34   
Willowdale, Willowdale West                                           4   
Woburn                                                                4   
Woodbine Heights                                                      7   
York Mills West                                                       3   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                            4               4   
Alderwood, Long Branch                               7               7   
Bathurst Manor, Wilson Heights, Downsview North     22              22   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   21              21   
...                                                ...             ...   
Willowdale, Willowdale East                         34              34   
Willowdale, Willowdale West                          4               4   
Woburn                                               4               4   
Woodbine Heights                                     7               7   
York Mills West                                      3               3   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                      4   
Alderwood, Long Branch                                         7   
Bathurst Manor, Wilson Heights, Downsview North               22   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             21   
...                                                          ...   
Willowdale, Willowdale East                                   34   
Willowdale, Willowdale West                                    4   
Woburn                                                         4   
Woodbine Heights                                               7   
York Mills West                                                3   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                     4  
Alderwood, Long Branch                                        7 

In [44]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2         Parkwoods                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Truck Stop  \
0                    0             0  ...           0   
1                    0             0  ...           0   
2                    0             0  ...           0   
3                    0             0  ...           0   
4                    0             0  ...           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 271 columns]

In [45]:
toronto_onehot.shape

(2122, 271)

In [46]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Truck Stop  \
0               0.0             0.000000           0.0  ...         0.0   
1               0.0             0.000000           0.0  ...         0.0   
2               0.0             0.000000           0.0  ...         0.0   
3               0.0             0.000000           0.0  ...         0.0   
4               0.0             0.047619           0.0  ...         0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 271 columns]

In [47]:
toronto_grouped.shape

(95, 271)

In [48]:

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [49]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt          Skating Rink   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0                Lounge        Breakfast Spot  Latin American Restaurant   
1                   Pub                   Gym             Sandwich Place   
2           Coffee Shop        Sandwich Place                      Diner   
3   Japanese Restaurant                  Bank         Chinese Restaurant   
4        Sandwich Place           Coffee Shop              Grocery Store   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Eastern European Restaurant   Distribution Center               Dog Run   
1                  Coffee Shop          Skating Rink    Dim Sum Restaurant   
2                   Restaurant      Sushi Restaurant         Deli / Bodega   
3                        Diner        Discount Store   Distribution Center   
4                    Juice Bar                  Café     Indian Restaurant   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Doner Restaurant            Donut Shop              Drugstore  
1                      Diner        Discount Store    Distribution Center  
2  Middle Eastern Restaurant           Bridal Shop         Ice Cream Shop  
3                    Dog Run      Doner Restaurant            Yoga Studio  
4                 Restaurant               Butcher                    Pub

In [50]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([4, 0, 4, 4, 4, 4, 4, 4, 4, 4], dtype=int32)

In [62]:
# add clustering labels
#neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = combined_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             1.0     Food & Drink Shop   
1  43.725882 -79.315572             4.0          Intersection   
2  43.654260 -79.360636             4.0           Coffee Shop   
3  43.718518 -79.464763             4.0        Clothing Store   
4  43.662301 -79.389494             4.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                  Pool             Drugstore   
1     French Restaurant           Coffee Shop          Hockey Arena   
2                Bakery                  Park                  Café   
3     Accessories Store           Coffee Shop           Event Space   
4     College Cafeteria      Sushi Restaurant           Yoga Studio   

    5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0                   Diner        Discount Store    Distribution Center   
1   Portuguese Restaurant            Donut Shop                  Diner   
2                     Pub               Theater         Breakfast Spot   
3  Furniture / Home Store              Boutique  Vietnamese Restaurant   
4               Nightclub              Beer Bar          Smoothie Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dog Run      Doner Restaurant             Donut Shop  
1        Discount Store   Distribution Center                Dog Run  
2            Beer Store        Chocolate Shop      French Restaurant  
3     Convenience Store                 Diner     Falafel Restaurant  
4         Burrito Place        Sandwich Place                   Café

In [63]:
toronto_merged[toronto_merged['Cluster Labels'].isnull()]

Postal Code      Borough  \
5          M9A    Etobicoke   
11         M9B    Etobicoke   
45         M2L   North York   
95         M1X  Scarborough   

                                        Neighbourhood   Latitude  Longitude  \
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242   
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724   
45                           York Mills, Silver Hills  43.757490 -79.374714   
95                                        Upper Rouge  43.836125 -79.205636   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
5              NaN                   NaN                   NaN   
11             NaN                   NaN                   NaN   
45             NaN                   NaN                   NaN   
95             NaN                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
5                    NaN                   NaN                   NaN   
11                   NaN                   NaN                   NaN   
45                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
5                    NaN                   NaN                   NaN   
11                   NaN                   NaN                   NaN   
45                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   9th Most Common Venue 10th Most Common Venue  
5                    NaN                    NaN  
11                   NaN                    NaN  
45                   NaN                    NaN  
95                   NaN                    NaN

In [64]:

import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Cluster 1

In [65]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8         East York             0.0           Pizza Place   
10       North York             0.0           Pizza Place   
32      Scarborough             0.0           Pizza Place   
50       North York             0.0           Pizza Place   
51      Scarborough             0.0   American Restaurant   
63             York             0.0           Pizza Place   
67  Central Toronto             0.0           Pizza Place   
70        Etobicoke             0.0           Pizza Place   
72       North York             0.0           Pizza Place   
77        Etobicoke             0.0           Pizza Place   
82      Scarborough             0.0           Pizza Place   
89        Etobicoke             0.0         Grocery Store   
93        Etobicoke             0.0           Pizza Place   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8     Athletics & Sports             Gastropub          Intersection   
10   Japanese Restaurant                Bakery                   Pub   
32            Playground             Drugstore                 Diner   
50                   Gym          Intersection            Donut Shop   
51                 Motel    Falafel Restaurant           Event Space   
63        Breakfast Spot     Convenience Store             Drugstore   
67        Breakfast Spot                   Gym                 Hotel   
70           Coffee Shop        Sandwich Place        Discount Store   
72         Grocery Store           Coffee Shop              Pharmacy   
77                  Park              Bus Line        Sandwich Place   
82              Pharmacy                  Bank          Intersection   
89           Pizza Place   Fried Chicken Joint        Sandwich Place   
93                   Pub                   Gym        Sandwich Place   

      5th Most Common Venue 6th Most Common Venue  \
8               Flea Market             Pet Store   
10              Escape Room     Electronics Store   
32           Discount Store   Distribution Center   
50       Dim Sum Restaurant                 Diner   
51     Ethiopian Restaurant           Escape Room   
63                    Diner        Discount Store   
67        Food & Drink Shop      Department Store   
70       Chinese Restaurant          Intersection   
72  Comfort Food Restaurant    Dim Sum Restaurant   
77              Escape Room     Electronics Store   
82       Italian Restaurant       Thai Restaurant   
89           Discount Store            Beer Store   
93              Coffee Shop          Skating Rink   

          7th Most Common Venue        8th Most Common Venue  \
8                      Pharmacy                         Café   
10  Eastern European Restaurant                    Drugstore   
32                      Dog Run             Doner Restaurant   
50               Discount Store          Distribution Center   
51            Electronics Store  Eastern European Restaurant   
63          Distribution Center                      Dog Run   
67                 Dance Studio                         Park   
70    Middle Eastern Restaurant                    Drugstore   
72           Falafel Restaurant                  Event Space   
77  Eastern European Restaurant                    Drugstore   
82           Chinese Restaurant                  Gas Station   
89          Japanese Restaurant         Fast Food Restaurant   
93           Dim Sum Restaurant                        Diner   

   9th Most Common Venue       10th Most Common Venue  
8         Breakfast Spot                         Bank  
10            Donut Shop                 Dessert Shop  
32            Donut Shop  Eastern European Restaurant  
50               Dog Run             Doner Restaurant  
51    Dim Sum Restaurant                    Drugstore  
63      Doner Restaurant                   Donut Shop  
67        Sandwich Place         Gym / Fitness Center  
70            Donut Shop  Eastern European Restaurant  
72  

Cluster 2

In [66]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             1.0     Food & Drink Shop   
21              York             1.0                  Park   
35         East York             1.0                  Park   
49        North York             1.0                Bakery   
52        North York             1.0                   Gym   
61   Central Toronto             1.0                  Park   
66        North York             1.0                  Park   
68   Central Toronto             1.0                  Park   
83   Central Toronto             1.0          Tennis Court   
85       Scarborough             1.0            Playground   
91  Downtown Toronto             1.0                  Park   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
0                   Park                        Pool             Drugstore   
21         Women's Store                        Pool             Drugstore   
35     Convenience Store                Intersection           Yoga Studio   
49                  Park  Construction & Landscaping           Yoga Studio   
52                  Park                 Yoga Studio             Drugstore   
61              Bus Line                 Swim School           Yoga Studio   
66                   Bar           Convenience Store           Yoga Studio   
68      Sushi Restaurant               Jewelry Store                 Trail   
83                 Trail                        Park            Playground   
85                  Park                Intersection           Yoga Studio   
91                 Trail                  Playground           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Diner        Discount Store   Distribution Center   
21                 Diner        Discount Store   Distribution Center   
35            Donut Shop                 Diner        Discount Store   
49             Drugstore        Discount Store   Distribution Center   
52                 Diner        Discount Store   Distribution Center   
61            Donut Shop        Discount Store   Distribution Center   
66             Drugstore        Discount Store   Distribution Center   
68           Yoga Studio            Donut Shop        Discount Store   
83    Dim Sum Restaurant                 Diner        Discount Store   
85      Doner Restaurant    Dim Sum Restaurant                 Diner   
91    Dim Sum Restaurant                 Diner        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Dog Run      Doner Restaurant             Donut Shop  
21               Dog Run      Doner Restaurant             Donut Shop  
35   Distribution Center               Dog Run       Doner Restaurant  
49               Dog Run      Doner Restaurant             Donut Shop  
52               Dog Run      Doner Restaurant             Donut Shop  
61               Dog Run      Doner Restaurant              Drugstore  
66               Dog Run      Doner Restaurant             Donut Shop  
68   Distribution Center               Dog Run       Doner Restaurant  
83   Distribution Center               Dog Run       Doner Restaurant  
85        Discount Store   Distribution Center                Dog Run  
91   Distribution Center               Dog Run       Doner Restaurant

Cluster 3

In [67]:

toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             2.0  Fast Food Restaurant    Dim Sum Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6        Farmers Market    Falafel Restaurant           Event Space   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6  Ethiopian Restaurant           Escape Room     Electronics Store   

         9th Most Common Venue 10th Most Common Venue  
6  Eastern European Restaurant              Drugstore

Cluster 4

In [68]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             3.0        Baseball Field           Yoga Studio   
101   Etobicoke             3.0                  Pool        Baseball Field   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57              Drugstore        Discount Store   Distribution Center   
101           Yoga Studio             Drugstore        Discount Store   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57                Dog Run      Doner Restaurant            Donut Shop   
101   Distribution Center               Dog Run      Doner Restaurant   

           9th Most Common Venue       10th Most Common Venue  
57   Eastern European Restaurant                        Field  
101                   Donut Shop  Eastern European Restaurant

Cluster 5

In [70]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             4.0          Intersection   
2    Downtown Toronto             4.0           Coffee Shop   
3          North York             4.0        Clothing Store   
4    Downtown Toronto             4.0           Coffee Shop   
7          North York             4.0                   Gym   
..                ...             ...                   ...   
97   Downtown Toronto             4.0           Coffee Shop   
98          Etobicoke             4.0                 River   
99   Downtown Toronto             4.0           Coffee Shop   
100      East Toronto             4.0    Light Rail Station   
102         Etobicoke             4.0        Hardware Store   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1         French Restaurant           Coffee Shop          Hockey Arena   
2                    Bakery                  Park                  Café   
3         Accessories Store           Coffee Shop           Event Space   
4         College Cafeteria      Sushi Restaurant           Yoga Studio   
7                Beer Store   Japanese Restaurant           Coffee Shop   
..                      ...                   ...                   ...   
97                     Café                 Hotel                   Gym   
98                     Pool           Yoga Studio      Doner Restaurant   
99      Japanese Restaurant      Sushi Restaurant               Gay Bar   
100             Yoga Studio         Auto Workshop         Garden Center   
102  Thrift / Vintage Store            Kids Store  Fast Food Restaurant   

      5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
1     Portuguese Restaurant            Donut Shop                  Diner   
2                       Pub               Theater         Breakfast Spot   
3    Furniture / Home Store              Boutique  Vietnamese Restaurant   
4                 Nightclub              Beer Bar          Smoothie Shop   
7                Restaurant    Dim Sum Restaurant         Sandwich Place   
..                      ...                   ...                    ...   
97               Restaurant   Japanese Restaurant             Steakhouse   
98       Dim Sum Restaurant                 Diner         Discount Store   
99               Restaurant  Fast Food Restaurant            Yoga Studio   
100                  Garden  Fast Food Restaurant         Farmers Market   
102          Discount Store     Convenience Store                    Gym   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1          Discount Store   Distribution Center                Dog Run  
2              Beer Store        Chocolate Shop      French Restaurant  
3       Convenience Store                 Diner     Falafel Restaurant  
4           Burrito Place        Sandwich Place                   Café  
7               Bike Shop   Sporting Goods Shop                   Café  
..                    ...                   ...                    ...  
97     Seafood Restaurant   American Restaurant          Deli / Bodega  
98    Distribution Center               Dog Run             Donut Shop  
99                    Pub                  Café            Men's Store  
100            Comic Shop                  Park   Gym / Fitness Center  
102        Sandwich Place          Burger Joint            Social Club  

[72 rows x 12 columns]